libpod/image/image.go

```diff
@@ -1076,3 +1078,144 @@ func (i *Image) Comment(ctx context.Context, manifestType string) (string, error
 	}
 	return ociv1Img.History[0].Comment, nil
 }
+
+type TreeImage struct {
+	ID       string
+	ParentID string
+	RepoTags []string
+	Size     int64
+}
+
+// Tree gets dependency list in tree format of the image
+func (i *Image) Tree(ctx context.Context, imageName string) (*bytes.Buffer, error) {
+
+	layerMap, err := i.layerToImageMap(ctx)
+	if err != nil {
+		return nil, err
+	}
+	layerMap, err = i.addImageToLayerInfo(ctx, layerMap)
+
+	repoTag, err := i.MatchRepoTag(imageName)
+	if err != nil {
+		return nil, err
+	}
+	// Get the image top layer and then parse as per parent
+	layer := i.TopLayer()
+	dependencyList := returnDependencyTree(layerMap, repoTag, layer)
+
+	var buffer bytes.Buffer
+	printTree(&buffer, dependencyList)
+
+	return &buffer, err
+}
+
+func printTree(buffer *bytes.Buffer, treeStack []TreeImage) {
+
+	prefixSpace := 0
+	for i := len(treeStack) - 1; i >= 0; i-- {
+		for p := 0; p < prefixSpace; p++ {
+			buffer.WriteString(fmt.Sprintf(" "))
+		}
+		buffer.WriteString(fmt.Sprintf("└─ ID: %s  Tags: %s VirtualSize: %v\n",
+			treeStack[i].ID[:12],
+			treeStack[i].RepoTags,
+			units.HumanSizeWithPrecision(float64(treeStack[i].Size), 4)),
+		)
+
+		prefixSpace++
+	}
+}
+
+func (i *Image) layerToImageMap(ctx context.Context) (map[string]*TreeImage, error) {
+	layerMap := make(map[string]*TreeImage)
+	layers, _ := i.imageruntime.store.Layers()
+	for _, layer := range layers {
+		existingLayer, ok := layerMap[layer.ID]
+		if !ok {
+			layerMap[layer.ID] = &TreeImage{
+				ID:       layer.ID,
+				Size:     layer.CompressedSize,
+				ParentID: layer.Parent,
+			}
+		} else {
+			return nil, fmt.Errorf("duplicate %s layer found in image store", existingLayer.ID)
+		}
+	}
+	return layerMap, nil
+}
+
+func (i *Image) addImageToLayerInfo(ctx context.Context, layerMap map[string]*TreeImage) (map[string]*TreeImage, error) {
+	images, err := i.imageruntime.GetImages()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting images from store")
+	}
+
+	for _, image := range images {
+
+		layerid := image.TopLayer()
+		imgLayer, err := image.imageruntime.store.Layer(layerid)
+		if err != nil {
+			return nil, err
+		}
+
+		for imgLayer != nil {
+			existingLayer, ok := layerMap[imgLayer.ID]
+			if !ok {
+				return nil, fmt.Errorf("layer %s not found ", imgLayer.ID)
+			} else {
+				names := image.Names()
+				existingLayer.RepoTags = append(existingLayer.RepoTags, names...)
+			}
+			imgLayer = image.getLayer(imgLayer.Parent)
+		}
+	}
+	return layerMap, nil
+}
+
+func (i *Image) getLayer(layerID string) *storage.Layer {
+	layers, err := i.imageruntime.store.Layers()
+	if err != nil {
+		return nil
+	}
+
+	for _, layer := range layers {
```

Comment:
This is O(layers) and all of this function should be unnecessary because the caller already has access to the built `layerMap` and can just index by ID.